File: mobile/src/main/java/com/siliconlabs/bledemo/adapters/DeviceInfoViewHolder.java
package com.siliconlabs.bledemo.adapters;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.siliconlabs.bledemo.ble.BluetoothDeviceInfo;

public abstract class DeviceInfoViewHolder<T extends BluetoothDeviceInfo> extends RecyclerView.ViewHolder implements View.OnClickListener {

    View.OnClickListener onClickListener;

    public abstract void setData(T info, int position);

    public DeviceInfoViewHolder(View itemView) {
        super(itemView);
    }

    @Override
    public void onClick(View v) {
        if (onClickListener != null) {
            onClickListener.onClick(v);
        }
    }

    public void setOnClickListener(View.OnClickListener onClickListener) {
        this.onClickListener = onClickListener;
    }

    public abstract static class Generator {

        private int layoutResId;

        public Generator(int resId) {
            this.layoutResId = resId;
        }

        public DeviceInfoViewHolder generate(ViewGroup parent) {
            final LayoutInflater li = LayoutInflater.from(parent.getContext());
            View itemView = li.inflate(layoutResId, parent, false);
            return generate(itemView);
        }

        public abstract DeviceInfoViewHolder generate(View itemView);
    }

}

File: mobile/src/main/java/com/siliconlabs/bledemo/Adapters/DeviceInfoViewHolder.kt
package com.siliconlabs.bledemo.Adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.siliconlabs.bledemo.Bluetooth.BLE.BluetoothDeviceInfo

abstract class DeviceInfoViewHolder(itemView: View?) : RecyclerView.ViewHolder(itemView!!), View.OnClickListener {
    private var onClickListener: View.OnClickListener? = null

    abstract fun setData(info: BluetoothDeviceInfo, position: Int, size: Int)
    override fun onClick(v: View) {
        onClickListener?.onClick(v)
    }

    fun setOnClickListener(onClickListener: View.OnClickListener?) {
        this.onClickListener = onClickListener
    }

    abstract class Generator(private val layoutResId: Int) {
        fun generate(parent: ViewGroup): DeviceInfoViewHolder {
            val li = LayoutInflater.from(parent.context)
            val itemView = li.inflate(layoutResId, parent, false)
            return generate(itemView)
        }

        abstract fun generate(itemView: View): DeviceInfoViewHolder

    }
}